File: Largest-Rectangle-in-Histogram/Largest-Rectangle-in-Histogram.py


class Solution:
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height):
        size=len(height)
        if size==0:
            return 0
        sum=0
        s=[]
        height.append(0)
        size+=1
        i=0
        while i<size:
            if len(s)==0 or height[i]>height[s[-1]]:
                s.append(i)
            else:
                temp=s[-1]
                s.pop(-1)
                sum=max(sum,height[temp]*(i if len(s)==0 else i-s[-1]-1))
                i=i-1
            i=i+1
        return sum
        
 

File: solutions/Largest-Rectangle-in-Histogram/Largest-Rectangle-in-Histogram.cpp

// class Solution {
// public:
//     int largestRectangleArea(vector<int> &height) {
//         stack<int> st;
//         height.push_back(0);
//         int n=height.size();
//         int ans=0;
//         for(int i=0;i<n;i++)
//         {
//             if(st.empty() || height[i]>height[st.top()])
//                 st.push(i);
//             else
//             {
//                 int index=st.top();
//                 st.pop();
//                 ans=max(ans,height[index]*(st.empty()?i:i-st.top()-1));
//                 i--;
//             }
//         }
//         return ans;
//     }
// };

// Review - 20190726
// O(n)
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> st;
        heights.push_back(0);
        int ans = 0;
        for(int i=0;i<heights.size();i++)
        {
            if(st.empty() || heights[i] > heights[st.top()])
            {
                st.push(i);
            }
            else
            {
                int top = st.top();
                st.pop();
                ans = max(ans, heights[top] * (st.empty() ? i : i - st.top() - 1));
                i--;
            }
        }
        return ans;
    }
};
 